cpu/samd5x/periph/can.c

```diff
@@ -0,0 +1,768 @@
+/*
+ * Copyright (C) 2023 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_samd5x
+ * @{
+ *
+ * @file
+ * @brief       Implementation of the CAN controller driver
+ *
+ * @author      Firas Hamdi <firas.hamdi@ml-pa.com>
+ * @}
+ */
+
+#include <assert.h>
+#include <string.h>
+
+#include "periph/can.h"
+#include "periph/gpio.h"
+#include "can/device.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#define CANDEV_SAMD5X_CLASSIC_FILTER 0x02 /* Value from SAMD5x/E5x Family datasheet, Tables 39-13 and 39-17 */
+#define CANDEV_SAMD5X_INTERNAL_LOOPBACK 0    /* Set to 1 to access the internal loopback mode. Check SAMD5x/E5x Family datasheet, Figure 39-4 */
+
+/**
+ * @brief Specific configuration of the CAN filter
+ */
+enum {
+    CANDEV_SAMD5X_FILTER_DISABLE = 0x00,
+    CANDEV_SAMD5X_FILTER_RX_FIFO_0,
+    CANDEV_SAMD5X_FILTER_RX_FIFO_1
+};
+
+/**
+ * @brief Configuration of how to handles frames not matching the CAN filters
+ */
+enum {
+    CAN_ACCEPT_RX_FIFO_0 = 0x00, /* Direct frames not matching any CAN filters applied to Rx FIFO 0 */
+    CAN_ACCEPT_RX_FIFO_1, /* Direct frames not matching any CAN filters applied to Rx FIFO 1 */
+    CAN_REJECT /* Reject all frames not matching any CAN filters applied */
+};
+
+typedef enum {
+    MODE_INIT,
+    MODE_TEST,
+} can_mode_t;
+
+typedef struct __attribute__((packed)) {
+    uint8_t put:4;          /** Message Marker Put index */
+    uint8_t get:4;          /** Message Marker Get index */
+} can_mm_t;
```

Comment:
there is no reason to pack this  

i checked the code where it is used, the only place where the packed come into play is probaly inlined 

if that wouldn't be the case the packing would greatly reduce the performance when accessing 

you should remove the `__attribute__((packed))`